torch/nn/utils/prune.py

```diff
@@ -0,0 +1,1269 @@
+        >>> print(m.bias_mask)
+        tensor([1., 1., 1.])
+    """
+    IdentityPruningMethod.apply(module, name)
+    return module
+
+
+def random_unstructured(module, name, amount):
+    r"""Prunes tensor corresponding to parameter called `name` in `module`
+    by removing the specified `amount` of units selected at random.
+    Modifies module in place (and also return the modified module) 
+    by:
+    1) adding a named buffer called `name+'_mask'` corresponding to the 
+    binary mask applied to the parameter `name` by the pruning method.
+    2) replacing the parameter `name` by its pruned version, while the
+    original (unpruned) parameter is stored in a new parameter named 
+    `name+'_orig'`.
+
+    Args:
+        module (nn.Module): module containing the tensor to prune
+        name (string): parameter name within `module` on which pruning
+                will act.
+        amount (int or float): quantity of parameters to prune.
+            If float, should be between 0.0 and 1.0 and represent the
+            fraction of parameters to prune. If int, it represents the 
+            absolute number of parameters to prune.
+
+    Returns:
+        module (nn.Module): modified (i.e. pruned) version of the input
+            module.
+
+    Examples:
+        >>> m = prune.random_unstructured(nn.Linear(2, 3), 'weight', amount=1)
+        >>> torch.sum(m.weight_mask == 0)
+        tensor(1)
+
+    """
+    RandomPruningMethod.apply(module, name, amount)
+    return module
+
+
+def l1_unstructured(module, name, amount):
+    r"""Prunes tensor corresponding to parameter called `name` in `module`
+    by removing the specified `amount` of units with the lowest L1-norm.
+    Modifies module in place (and also return the modified module) 
+    by:
+    1) adding a named buffer called `name+'_mask'` corresponding to the 
+    binary mask applied to the parameter `name` by the pruning method.
+    2) replacing the parameter `name` by its pruned version, while the 
+    original (unpruned) parameter is stored in a new parameter named 
+    `name+'_orig'`.
+
+    Args:
+        module (nn.Module): module containing the tensor to prune
+        name (string): parameter name within `module` on which pruning
+                will act.
+        amount (int or float): quantity of parameters to prune.
+            If float, should be between 0.0 and 1.0 and represent the
+            fraction of parameters to prune. If int, it represents the 
+            absolute number of parameters to prune.
+
+    Returns:
+        module (nn.Module): modified (i.e. pruned) version of the input
+            module.
+
+    Examples:
+        >>> m = prune.l1_unstructured(nn.Linear(2, 3), 'weight', amount=0.2)
+        >>> m.state_dict().keys()
+        odict_keys(['bias', 'weight_orig', 'weight_mask'])
+    """
+    L1PruningMethod.apply(module, name, amount)
+    return module
+
+
+def random_structured(module, name, amount, axis):
```

Comment:
while all of these methods were exposed for taking in a `module` and `name`, what about if we want to use these methods functionally, directly on a `weight` Tensor?

For example:

```python
weight = torch.randn(20, 30)
weight_pruned = prune.random_structured(weight, amount=0.4, axis=1)
```

Reply:
as of now, these three lines will do it:
```python
weight = torch.randn(20, 30)
rsp = prune.RandomStructuredPruningMethod(amount=0.4, dim=1)
pruned_weight = weight * rsp.compute_mask(weight, default_mask=torch.ones_like(weight))
```

Reply:
honestly, doesn't look great.
Can we check if `Module` or `Tensor` and pack the `Tensor` if needed?

I asked the question because I suspect people will want to do it more often than you think.

Reply:
hmmk, so what's your suggested api here? what does "pack the Tensor" mean?

Reply:
first, I think `prune.RandomStructuredPruningMethod` is sooo much more verbosity than `prune.random_structured`. Can we have them align better. For example the class be called: `prune.RandomStructured`.

Second, I was wondering if something like: `pruned_weight = rsp.prune(weight)` that implicitly will do `weight * rsp.compute_mask(weight, default_mask=torch.ones_like(weight))`

Reply:
I've changed them all to align with the function names and added a `.prune` method to the base pruning class. I added a quick test too.